draft-regext-rdap-extensions.md

```diff
@@ -379,6 +381,59 @@ extension registration that attempts to use one of the RDAP
 conformance values given in this section as an extension identifier
 (and so as an RDAP conformance value also) will be rejected.
 
+# Redirects {#redirects}
+
+[@!RFC7480] describes the use of redirects in RDAP. Redirects are prominent
+in the discovery of authoritative INR servers as the process outlined in
+[@?RFC9224], which uses IANA allocations, does not account for transfers of
+resources between INRs. [@!RFC7480, Section 4.3] instructs servers to ignore
+unknown query parameters. As it relates to issuing URLs for redirects, servers
+MUST NOT blindly copy query parameters from a request to a redirect URL as
+query parameters may contain sensitive information, such as security credentials,
+not relevant to the target server of the URL. Following the advice in [@!RFC7480],
+servers SHOULD only place query paramters in redirect URLs when it is known
+by the origin server (the server issuing the redirect) that the target server
+(the server referenced by the redirect) can process the query parameter and the
+contents of the query parameter are appropriate to be received by the target.
+
+As it is unlikely that every server in a cross-authority, redirect scenario will be upgraded to process
+every new extension, extensions should not solely rely on query parameters to convey
+information about a resource as query parameters are not guaranteed to survive a redirect.
+
+This does not mean extensions are prohibited from using query parameters, but
+rather that the use of query parameters must be applied for the scenarios appropriate
+for the use of the extension. Therefore, extensions MUST NOT rely on query paramters
+when the extension is to be used in scenarios requiring clients to find authoritative
+servers, such as that described above, or other scenarios using redirects among 
+servers of differing authorities.
```

Comment:
I think an extension can rely on a query parameter in a redirect scenario so long as clients are aware that they have to add those parameters to the redirect URLs that they receive from servers, so I think advice like "consider whether relying on query parameters increases the chance of client error" might be better than a blanket MUST NOT.

Reply:
Again, placing the complexity on the client is not good. It is far easier for a client to use rdap-x than to have to handle redirect loops, query parameter conflicts, and other issues. This also directly conflicts with RFC 7480 section 5.2 which says clients shouldn't have to transform redirect URLs. Forcing clients to directly handle all the issues with redirects makes using tools like `curl` with RDAP very difficult.

Reply:
I added a paragraph on client complexity.

Reply:
> Again, placing the complexity on the client is not good. It is far easier for a client to use rdap-x than to have to handle redirect loops, query parameter conflicts, and other issues.

I think using RDAP-X makes sense for clients where adding/setting headers is easy, but I wonder if leaving the query string option open makes life a bit easier for working with RDAP in e.g. the browser.

How does RDAP-X help to avoid handling redirect loops?

> This also directly conflicts with RFC 7480 section 5.2 which says clients shouldn't have to transform redirect URLs.

That's a fair point, but how does that square with the earlier 'SHOULD' around setting query parameters in redirect URLs?  Presumably redirect servers 'MUST' add (safe) query parameters that exist in the original query to any redirect URLs that they return, in order to comply with this requirement from 7480.

Reply:
> I think using RDAP-X makes sense for clients where adding/setting headers is easy, but I wonder if leaving the query string option open makes life a bit easier for working with RDAP in e.g. the browser.

Well, RDAP-X is only good for signalling extensions... it isn't the answer to all problems. So I think you are right that an outright prohibition is not good.

Browsers can set media types in JavaScript and WASM.

> How does RDAP-X help to avoid handling redirect loops?

They still need to be handled, but the library will do that in almost all cases. Therefore, the client library is just told to use the media type and the client library does the heavy lifting. We have sample code in the RDAP-X draft.

> That's a fair point, but how does that square with the earlier 'SHOULD' around setting query parameters in redirect URLs? Presumably redirect servers 'MUST' add (safe) query parameters that exist in the original query to any redirect URLs that they return, in order to comply with this requirement from 7480.

A server SHOULD only set query parameters in redirect URLs when the target is the proper recipient, and the client does not need to do anything to that URL, just as 7480 says.

Reply:
> They still need to be handled, but the library will do that in almost all cases. Therefore, the client library is just told to use the media type and the client library does the heavy lifting. We have sample code in the RDAP-X draft.

Ah yep, I get it now, thanks (was thinking of 'loops' in the sense of infinite redirect loops).